src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -750,9 +750,9 @@ private static string GetMatchingFrameworkStrings(PackageSpec spec, NuGetFramewo
 
             foreach (var item in items)
             {
-                var dependencies =
-                    item.Data?.Dependencies?.Select(
-                        dependency => new PackageDependency(dependency.Name, VersionRange.All));
+                IEnumerable<PackageDependency> dependencies = item.Data?.Dependencies?
```

Comment:
Isn't LibraryDependencyReferenceType.Transitive the transitively pinned dependency? And none would be a transitive dependency that wasn't pinned?
Should we look at the direct ones instead?

Reply:
From my understanding, when you have 100 `<PackageVersion />` items and transitive pinning enabled, NuGet adds all of them as a dependency with `LibraryDependencyReferenceType.None`.  It then walk the graph and ones that actually take part in the tree get moved to `LibraryDependencyReferenceType.Direct` when pinning is enabled.  The ones with `LibraryDependencyReferenceType.None` are left in the list for later use but we're starting to add more and more places where we ignore them.

https://github.com/NuGet/NuGet.Client/blob/95d72977b8ed9f26a897fe1b048ac810495211c0/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L316

https://github.com/NuGet/NuGet.Client/blob/95d72977b8ed9f26a897fe1b048ac810495211c0/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L507

https://github.com/NuGet/NuGet.Client/pull/4950/files#diff-cf026661b18af5f436cbd15343ef7eaff39576ec89b7dc3c7df62fa0b6135d39R145

I'm also wondering if we need to look into removing these items so we don't have a long tail of bugs arising from not ignoring them somewhere else in the code.

Reply:
My main concern is that the enum has 3, so technically != None and == None are not true opposites.

Reply:
I think we need ordering for direct and transitive dependencies but can ignore these other ones.  I can't seem to find any tests for this ordering logic, do you happen to know where some might be?

Reply:
https://github.com/NuGet/NuGet.Client/blob/95d72977b8ed9f26a897fe1b048ac810495211c0/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs#L463 is the logic.

No clue where the tests are. 
Remove that line and see which test fails? :D (hopefully)

Reply:
Yeah so there are tests for the `TopologicalSortUtility.SortPackagesByDependencyOrder()` method but none for the logic that creates `PackageDependencyInfo` objects from the restore graph.  If I could repro this with a simple test I would gladly add a test but I've only been able to repro it by cloning and restoring [this change](https://dev.azure.com/devdiv/DevDiv/_git/VS-Platform/pullrequest/438402).

Reply:
I didn't find any case where the `ReferenceType` was `LibraryDependencyReferenceType.Transitive` so I'll update the check to only look at `LibraryDependencyReferenceType.Direct`.